crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs

```diff
@@ -0,0 +1,494 @@
+//! means that the request must be idempotent or retrying might be unsafe.**
+
+use mappable_rc::Marc;
+use operation_id::OpId;
+use std::{
+    any::Any,
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    marker::PhantomData,
+    sync::Mutex,
+};
+use tokio::sync::watch::{self, Receiver, Sender};
+
+/// For coalescing to work correctly the key the key we use for lookup has to be simultaneously:
+/// - very unique: distinct code can't have the same key
+/// - very duplicated: the same code must use the same key
+///
+/// To achieve this we use a key pair:
+/// - `key` is the KV key, obviously distinct keys must originate distinct requests.
+/// - `op_id` is the [`TypeId`] of the [monomorphisation][mono] of the [`coalesce`][coalesce]
+///     function. No two closures or futures have the same [`TypeId`], hence, no two
+///     monomorphisations of the [`coalesce`] method can have the same id. Thus, using the
+///     [`TypeId`] we can guarantee that even if two operations try to work on the same `key`, they
+///     won't clober each other since they are produced by different closures.
+///
+/// mono: <https://en.wikipedia.org/wiki/Monomorphization>
+/// coalesce: [`RequestCoalescer::coalesce`]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct OperationKey {
+    key: String,
+    op_id: OpId,
+}
+
+/// The state of the coalesced calls.
+#[derive(Default)]
+pub struct RequestCoalescer {
+    inflight: Mutex<HashMap<OperationKey, Receiver<Option<CoalescedValue>>>>,
+    #[cfg(feature = "test-utils")]
+    shutting_down: std::sync::atomic::AtomicBool,
+}
+
+impl RequestCoalescer {
+    #[cfg(feature = "test-utils")]
+    pub async fn reset(&self) {
+        self.shutting_down
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+        let recievers = self
+            .inflight
+            .lock()
+            .unwrap()
+            .values()
+            .cloned()
+            .collect::<Vec<_>>();
+
+        for mut recv in recievers {
+            let _ = recv.wait_for(|x| x.is_some()).await;
+        }
+
+        self.inflight.lock().unwrap().clear();
+        self.shutting_down
+            .store(false, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    pub async fn coalesce<F, Fut, R, E>(
+        &self,
+        key: String,
+        executor: F,
+    ) -> Result<Option<Marc<R>>, Marc<E>>
+    where
+        R: Send + Sync + 'static,
+        E: Send + Sync + 'static,
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = Result<Option<Marc<R>>, E>>,
+    {
+        let result = loop {
+            #[cfg(feature = "test-utils")]
+            assert!(
+                !self.shutting_down.load(std::sync::atomic::Ordering::SeqCst),
+                "RequestCoalescer is shutting down"
+            );
+
+            let op_key = OperationKey {
+                op_id: operation_id::of::<F, Fut, R, E>(),
+                key: key.clone(),
+            };
```

Comment:
can the `op_key` variable be defined before the loop?

Reply:
That results in an additional clone when passing it to the ResponseSender